### PR TITLE
add node services URI and fields

### PR DESCRIPTION
**Pull request by fmigneault**
## Overview

Add node services URI and fields.

## Changes

**Non-breaking changes**
- Node Services: Add definitions and variables for every service represented by
  the [DACCS-Climate/Marble-node-registry](https://github.com/DACCS-Climate/Marble-node-registry).

  - Add `version` field using the corresponding `<SERVICE>_VERSION` variables.
  - Add `types` field restricted by specific values instead of previous `keywords` expected to be extendable.
  - Add `<SERVICE>_IMAGE_URI` variables to provide `rel: service-meta` link for every service.

**Breaking changes**
- n/a

## Related Issue / Discussion

- Resolves https://github.com/bird-house/birdhouse-deploy/issues/441
- Depends on https://github.com/DACCS-Climate/Marble-node-registry/pull/40 to enforce validation

## CI Operations

birdhouse_daccs_configs_branch: master
birdhouse_skip_ci: false

### CODE REVIEW

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [inline birdhouse/components/finch/service-config.json.template:8]: should we leave these keywords as is if we're moving this information to "types". What if we borrowed some of the tags and categories from canarie-api instead. So for finch we could do something like:

```
"keywords": ["Climatology", "Cloud", "Data Slicer", "Climate Indices", ...]
```

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [inline birdhouse/components/finch/service-config.json.template:2]: Let's update this to specify version 1.2.0 as well

Comment (mishaschwartz) [reply birdhouse/components/finch/service-config.json.template:2]: And then update the test in `tests/test_deployment.py` so that it gets the schema from the "$schema" keyword in the json file instead of using the `services_config_schema` pytest fixture.

That way, we can ensure that the schema is validated by the correct version of the schema each time.

File: birdhouse/components/raven/docker-compose-extra.yml
Comment (mishaschwartz) [inline birdhouse/components/raven/docker-compose-extra.yml:12]: I like that we're now being consistent with how we specify images instead of before where we had some hardcoded image names and some in variables from default.env. I like that this PR will make this more consistent

File: birdhouse/components/finch/service-config.json.template
Comment (tlvu) [inline birdhouse/components/finch/service-config.json.template:31]: If a new var is used for template expansion it should be added to `VARS` or `OPTIONAL_VARS` in the `default.env` of the corresponding component.  Otherwise, template expansion will not work.

I guess we do not validate the value in "href", that's why we did not caught the wrong template expansion?

**Review by tlvu — CHANGES_REQUESTED** [review]
Ensure template expansion actually work.

Also I do not think this PR introduce a breaking change (in reference to this PR description).  The new var has a default value so upgrading to this PR is 100% seamless, no manual intervention required, so nothing "breaking" as per deployment as I can see.

I defer to Misha for review of the content of the .json .

Agree with Misha this PR unified the way to specify all the images so it's a very nice side-effect.

Also, it seems all we need are additional copy/paste to make deprecated-components work so probably cheap enough to make it work as well.

**Comment by mishaschwartz** [comment]
@tlvu 

> Also, it seems all we need are additional copy/paste to make deprecated-components work so probably cheap enough to make it work as well.

Cheap enough for sure, but can we also plan to stop updating these at some point? I mean, at what point do we actually treat them as deprecated instead of just in another folder that happens to be called `deprecated-components`?

> I guess we do not validate the value in "href", that's why we did not caught the wrong template expansion?

We are, it's just that the [lnks schema](https://json-schema.org/draft/2020-12/links) uses the [uri-template](https://json-schema.org/understanding-json-schema/reference/string#uri-template) string type to check the `href` value and `${some_string_here}` is a valid uri-template according to [RFC6570](https://datatracker.ietf.org/doc/html/rfc6570). There's not much we can do about that.

**Comment by tlvu** [comment]
> at what point do we actually treat them as deprecated instead of just in another folder that happens to be called `deprecated-components`?

I guess at the point where it's expensive.  I agree "expensive" is very vague so up to you I guess.

File: birdhouse/components/finch/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/finch/service-config.json.template:8]: Yes. I would like to do this, but until there is a clear-cut version I can refer to, adding these to the keywords will fail schema validation.

**Comment by fmigneault** [comment]
I did omit `flyingpigeon` on purpose due to deprecated, but as highlighted, it is fairly easy to update, so I can add it as well.
Will wait for a release that integrates the relevant updates to the schema to update the references.

**Comment by fmigneault** [comment]
> We are, it's just that the [lnks schema](https://json-schema.org/draft/2020-12/links) uses the [uri-template](https://json-schema.org/understanding-json-schema/reference/string#uri-template) string type to check the href value and ${some_string_here} is a valid uri-template according to [RFC6570](https://datatracker.ietf.org/doc/html/rfc6570). There's not much we can do about that.

I don't think this is the cause. 
According to JSON-schema `format` is just a hint, not an enforced validation, unless otherwise specified.
```python
jsonschema.validate("^^^!!!! <<SO NOT AN URI>> ???", {"type": "string", "format": "uri-template"})  # OK
```
Looking up the format check in the validator confirms this (`uri` and `uri-template` not checked):
```python
jsonschema.draft202012_format_checker
<FormatChecker checkers=['date', 'email', 'idn-email', 'idn-hostname', 'ipv4', 'ipv6', 'regex', 'uuid']>
```
So, the real fix would be to add a `"pattern": "^https?://"` explicitly.

File: birdhouse/components/finch/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/finch/service-config.json.template:31]: Optional vars added. See other comments for the validation.

**Comment by mishaschwartz** [comment]
>  According to JSON-schema format is just a hint, not an enforced validation, unless otherwise specified.

Ok that's good to know....

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [reply birdhouse/components/finch/service-config.json.template:2]: Released! 

(https://github.com/DACCS-Climate/Marble-node-registry/releases/tag/1.2.0)

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [reply birdhouse/components/finch/service-config.json.template:8]: Got it. Try now with version [1.2.0](https://github.com/DACCS-Climate/Marble-node-registry/releases/tag/1.2.0)?

File: birdhouse/components/finch/default.env
Comment (tlvu) [inline birdhouse/components/finch/default.env:8]: I just realized that all xxx_IMAGE and xxx_URI vars should both be delayed eval because they use xxx_VERSION var and that can be overridden by `env.local`.

File: birdhouse/components/finch/default.env
Comment (tlvu) [inline birdhouse/components/finch/default.env:9]: In addition to be in `DELAYED_EVAL` list, all xxx_IMAGE and xxx_IMAGE_URL have to be in single quote for delayed eval to work properly.

File: birdhouse/components/finch/service-config.json.template
Comment (tlvu) [inline birdhouse/components/finch/service-config.json.template:31]: Just my 2 cents, since the URI, do not start with http, should we name the new field "uri" instead of "href"?  "href" would suggest starting with http usually.  Ignore if it has to be "href" for compat or other reasons.

Comment (mishaschwartz) [reply birdhouse/components/finch/service-config.json.template:31]: `"href"` is required by the schema so we can't change the name

File: birdhouse/components/finch/default.env
Comment (fmigneault) [reply birdhouse/components/finch/default.env:9]: Yes, I noticed after. In the process of rewriting with adjustment to unittests evaluating them as well.

File: birdhouse/components/finch/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/finch/service-config.json.template:31]: A non-http scheme value is a valid URI. It uniquely identifies the resource. 
I would prefer to have a `docker://` scheme or similar to be more explicit if permitted, but docker does not seem to like it. However, the values provided in the defaults can be used directly (e.g. `docker pull 'registry.hub.docker.com/pavics/weaver:5.0.0'` works, but `http://registry.hub.docker.com/pavics/weaver:5.0.0` doesn't), so I find them more useful although maybe counter-intuitive.

Maybe one way to avoid confusion would be to use `index.docker.io/pavics/weaver:5.0.0` instead, since it does not redirect to `https://hub.docker.com/` like `registry.hub.docker.com` does? This alternate location still woks with `docker pull 'index.docker.io/pavics/weaver:5.0.0'`, and it looks slightly more like the `type` definition. The idea of this link is really to have the exact value that can be plugged in the docker-compose service `image` field.

File: birdhouse/scripts/get-services-json.include.sh
Comment (mishaschwartz) [inline birdhouse/scripts/get-services-json.include.sh:18]: If you just left the stac/service-config.json.template file as is, all of this wouldn't be necessary.

I don't understand why this is needed

File: birdhouse/components/thredds/default.env
Comment (mishaschwartz) [inline birdhouse/components/thredds/default.env:6]: single quote

File: birdhouse/components/thredds/default.env
Comment (mishaschwartz) [inline birdhouse/components/thredds/default.env:7]: single quote

File: birdhouse/components/stac/service-config.json.template
Comment (mishaschwartz) [inline birdhouse/components/stac/service-config.json.template:1]: I don't understand why this change is necessary, it makes the format inconsistent with the rest of these files and makes parsing them more complicated (see other comment).

If the answer is that you want this file to be valid json ... I don't think that is a good enough reason to change this, this file is for internal use only and isn't meant to be read by any external client without being processed first.

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [reply birdhouse/components/finch/service-config.json.template:8]: Are we going to do this in this PR or in a different one?

File: birdhouse/components/finch/service-config.json.template
Comment (mishaschwartz) [inline birdhouse/components/finch/service-config.json.template:5]: This hasn't been implemented yet in the schema. Adding this won't cause schema validation to fail but do we plan on updating the schema before or after this PR is merged?

File: birdhouse/components/hummingbird/default.env
Comment (mishaschwartz) [inline birdhouse/components/hummingbird/default.env:39]: I don't think these are necessary to put here. It's not used in any templates

File: birdhouse/components/stac/default.env
Comment (tlvu) [inline birdhouse/components/stac/default.env:10]: Just curious, why not push the image with the actual tag `2.4.3-crim` instead of the generic tag `main`?

**Review by mishaschwartz — CHANGES_REQUESTED** [review]

File: birdhouse/components/stac/default.env
Comment (tlvu) [inline birdhouse/components/stac/default.env:17]: Same here, `3.0.0-beta.5-crim` instead of `docker_image_push`.  It is allowed to tag while on a branch.

**Review by tlvu — APPROVED** [review]
LGTM.  I have not done a deployment test to ensure all template expansions actually work.  I also defer to @mishaschwartz to validate the content of the various `.json` files and the test.

File: birdhouse/components/thredds/default.env
Comment (tlvu) [reply birdhouse/components/thredds/default.env:6]: Oh wow, Misha has much better eye than me this morning :D

File: birdhouse/components/stac/default.env
Comment (mishaschwartz) [reply birdhouse/components/stac/default.env:10]: See #371 and #346

Comment (fmigneault) [reply birdhouse/components/stac/default.env:10]: Yes, I could temporarily push a tag to make the reference more "semver"-like, but like @mishaschwartz highlighted, I would much rather have some work done on using the official image. The one tagged is already 9 months old. I don't want this patch tag to be considered a maintained revision.

File: birdhouse/scripts/get-services-json.include.sh
Comment (fmigneault) [reply birdhouse/scripts/get-services-json.include.sh:18]: Either this, or we patch everywhere that tries to load the JSON. 
I add errors when trying to extract the `$schema` in the tests.

File: birdhouse/components/finch/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/finch/service-config.json.template:5]: If the schema gets updated to include it beforehand, then we can bump its version right away. Otherwise, it can be done later. As mentioned, it doesn't cause an issue regardless at the moment since it is not validated.

File: birdhouse/components/finch/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/finch/service-config.json.template:8]: If possible, I'd prefer to work on this only once.

File: birdhouse/components/stac/service-config.json.template
Comment (fmigneault) [reply birdhouse/components/stac/service-config.json.template:1]: The format is also inconsistent currently. It is not a vaild JSON, so it cannot even be parsed unless hacking it before load.

Comment (fmigneault) [reply birdhouse/components/stac/service-config.json.template:1]: I prefer the all list approach since it works for every use case.

File: birdhouse/scripts/get-services-json.include.sh
Comment (fmigneault) [reply birdhouse/scripts/get-services-json.include.sh:18]: I purposely went without the `jq` approach to avoid introducing this extra dependency since the merge is relatively simple in this case.

One issue with the pseudo-JSON is that it is not obvious (for others) that the incorrectly formed content is intentional to work only once merged. Especially when working with IDEs, or investigating the generated JSON on the server, we get flagged about invalid JSON, which doesn't help maintain them. Since there is a way to make them valid JSON definitions and still work for the aggregation, I would favor using a valid list.

**Comment by fmigneault** [comment]
run tests

**Review by mishaschwartz — APPROVED** [review]
Looks good! Just double checking if there's still more keywords to add or not (see comment here: https://github.com/bird-house/birdhouse-deploy/pull/445#discussion_r1559968843)

**Comment by fmigneault** [comment]
For the moment, 1.2.0 still refers to `keywords` as only a specific set of values. Do you want to push a 1.3.0 right away with updated definitions allowing any value, and using `types` instead for the specific service type allowed?

**Comment by mishaschwartz** [comment]
Right, we're still discussing it here: https://github.com/DACCS-Climate/Marble-node-registry/issues/39

Ok nevermind then

**Pull request merged**
